relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,587 @@
+use crate::envelope::Envelope;
+use crate::extractors::StartTime;
+use crate::services::spooler::envelope_stack::EnvelopeStack;
+use futures::StreamExt;
+use relay_base_schema::project::ProjectKey;
+use sqlx::query::Query;
+use sqlx::sqlite::{SqliteArguments, SqliteRow};
+use sqlx::{Pool, QueryBuilder, Row, Sqlite};
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+use std::error::Error;
+use std::pin::pin;
+
+/// New Type used to define ordering on a `Box<Envelope>` based on the `start_time` field.
+struct OrderedEnvelope(Box<Envelope>);
+
+impl Eq for OrderedEnvelope {}
+
+impl PartialEq<Self> for OrderedEnvelope {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Equal)
+    }
+}
+
+impl PartialOrd<Self> for OrderedEnvelope {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for OrderedEnvelope {
+    fn cmp(&self, other: &Self) -> Ordering {
+        received_at(&other.0).cmp(&received_at(&self.0))
+    }
+}
+
+/// An error returned when doing an operation on [`SQLiteEnvelopeStack`].
+#[derive(Debug, thiserror::Error)]
+pub enum SQLiteEnvelopeStackError {
+    /// The stack is empty.
+    #[error("the stack is empty")]
+    Empty,
+
+    /// The database encountered an unexpected error.
+    #[error("a database error occurred")]
+    DatabaseError(#[from] sqlx::Error),
+}
+
+/// An [`EnvelopeStack`] that is implemented on an SQLite database.
+///
+/// For efficiency reasons, the implementation has an in-memory buffer that is periodically spooled
+/// to disk in a batched way.
+pub struct SQLiteEnvelopeStack {
+    db: Pool<Sqlite>,
+    spool_threshold: usize,
```

Comment:
This is subjective, but I would find it easier to define a `batch_size` here, and then derive the spool threshold as a multiple of the batch size.

Reply:
Definitely, I reasoned backwards when designing the thresholds, now it has been swapped.